web-app/js/smartR/_angular/controllers/heatmap.js

```diff
@@ -8,179 +8,115 @@ function HeatmapController ($scope, smartRUtils, commonWorkflowService) {
         },
-        subsets : 0,
-        scriptResults : {}
+        running: false,
+        subsets: 0,
+        scriptResults: {}
     };
 
     commonWorkflowService.registerCondition(
-        ['fetch.conceptBoxes.highDimensional'],
-        ['fetch.btn'],
-        this.whenSelectHighDimensionalNodes
-    );
-
-    commonWorkflowService.registerCondition(
-        ['fetch.btn.disabled', 'fetch.totalSamples', 'fetch.subsets'],
-        ['preprocess', 'runAnalysis'],
-        this.whenFetchHeatmapData
-    );
-
-    commonWorkflowService.registerCondition(
-        ['preprocess.btn.disabled'],
-        ['fetch', 'runAnalysis'],
-        this.whenPreprocessHeatmapData
-    );
-
-    commonWorkflowService.registerCondition(
-        ['runAnalysis.btn.disabled', 'runAnalysis.scriptResults'],
-        ['fetch', 'preprocess', 'runAnalysis.download'],
-        this.whenRunHeatmapAnalysis
+        ['fetch.running',
+            'preprocess.running',
+            'runAnalysis.running',
+            'fetch.conceptBoxes.highDimensional.valid'
+        ],
+        ['fetch', 'preprocess', 'runAnalysis'],
+        function(newValues, oldValues, scope, affectedComponents) {
+            var fetchRunning = newValues[0],
+                preprocessRunning = newValues[1],
+                runAnalysisRunning = newValues[2],
+                conceptBoxesHighDimensionalValid = newValues[3];
+            var fetchModel = affectedComponents[0],
+                preprocessModel = affectedComponents[1],
+                runAnalysisModel = affectedComponents[2];
+
+            // clear old results
+            if (fetchRunning) {
+                preprocessModel.scriptResults = {};
+                runAnalysisModel.scriptResults = {};
+            }
+
+            // clear old results
+            if (preprocessRunning) {
+                runAnalysisModel.scriptResults = {};
+            }
+
+            // disable tabs when certain criteria are not met
+            fetchModel.tab.disabled = preprocessRunning || runAnalysisRunning;
+            preprocessModel.tab.disabled = fetchRunning || runAnalysisRunning || !scope.fetch.loaded;
+                // FIXME: this is always zero. Probably has something to do with promises.
+                // || scope.fetch.totalSamples <= 1; 
+            runAnalysisModel.tab.disabled = fetchRunning || preprocessRunning || !scope.fetch.loaded;
+
+            // disable buttons when certain criteria are not met
+            fetchModel.btn.disabled = !conceptBoxesHighDimensionalValid;
+            runAnalysisModel.download.disabled = runAnalysisRunning || $.isEmptyObject(scope.runAnalysis.scriptResults);
+
+            // set ranking criteria
+            runAnalysisModel.subsets = scope.fetch.subsets;
+            if (scope.fetch.subsets < 2 &&
+                    ['logfold', 'bval', 'pval', 'adjpval', 'ttest'].indexOf(runAnalysisModel.params.ranking) !== -1) {
+                runAnalysisModel.params.ranking = 'mean';
```

Comment:
if subsets < 2 ---> rangking = 'coef'
if number of samples == 1 --> rangking = 'mean'